### PR TITLE
Add Union support

### LINKED ISSUE #24: Support fields with type typing.Union

### PR DESCRIPTION
**Pull request by daneoshiga**
Hi, firstly, thanks for the project :)

This is a first attempt of a feature proposal to add Union support, looking around the project I'm not sure if this is enough.

This way is possible to validate if the value is one of the ones available in the Union, and restrict to those.

it has to be done before any call to `issubclass(field_class` because that will fail with Union.

Hope this helps.

fixes #24

### CODE REVIEW

**Comment by eltonscm** [comment]
I think this also solves #23 because _"Optional[X] is equivalent to Union[X, None]"_ [0]. So, the following test should pass:

```python
@pytest.mark.parametrize('default', [1, None])
def test_model_validate_optional(default):
    class Foo(Model):
        optional: typing.Optional[int] = default

    foo = Foo()
    foo.validate()

    assert foo.optional == default
```

[0] https://docs.python.org/3/library/typing.html#typing.Optional

**Comment by lamenezes** [comment]
@daneoshiga thanks for your PR! I'll review it as soon as I can.

File: simple_model/fields.py [hunk 1]
```diff
@@ -37,6 +37,13 @@ def convert_to_type(self, instance, value, field_class=None):
         if not field_class or field_class is Any or value is None or self.is_property:
             return value
 
+        if field_class is Union:
+            assert any((type(value) is union_class for union_class in field_type.__args__)), (
```
Comment (lamenezes) [inline simple_model/fields.py]: maybe we can use instead:

```python
assert issubclass(type(value), field_type.__args__), ...
```

File: tests/test_fields.py [hunk 2]
```diff
@@ -128,6 +128,17 @@ def test_model_field_convert_to_type_unset(typeless_model_field):
...
 
 
+@pytest.mark.parametrize('value', (1, '1'))
+def test_model_field_convert_to_type_union(typeless_model_field, value):
+    assert typeless_model_field.convert_to_type(None, value, field_class=typing.Union[int, str]) is value
+
+
+def test_model_field_convert_to_type_union_invalid(typeless_model_field):
+    value = dict()
+    with pytest.raises(AssertionError):
```
Comment (lamenezes) [inline tests/test_fields.py]: I think we should check the message error just to be sure the wrong `AssertionError` isn't being captured

File: tests/test_models.py [hunk 3]
```diff
@@ -427,6 +427,16 @@ def test_field_factory_model():
...
 
 
+@pytest.mark.parametrize('default', ('1', 1, ))
+def test_model_validate_union(default):
+    class UnionModel(Model):
+        union: typing.Union[int, str] = default
+
+    union_model = UnionModel()
+    union_model.validate()
+    union_model.union == default
```
Comment (lamenezes) [inline tests/test_models.py]: I think this line is missing the `assert`

Comment (lamenezes) [reply tests/test_models.py]: also I believe the types should be checked

**Comment by lamenezes** [comment]
@daneoshiga thanks for your contribution! :soccer: :top:

**Comment by lamenezes** [comment]
@daneoshiga if you're interested the feature inserted in this PR can be further developed by supporting the cases below:

- field conversion considering all types of the `typing.Union`:
```python
class MyModel(Model):
  foo: Union[dict, str]

model = MyModel(foo={1: 2})
model.validate()  # ok

model2 = MyModel(foo=[(1, 2)])
model2.validate()  # no error
assert model2.foo == {1: 2}
```

- use `typing.Optional[str]` (which is exactly `typing.Union[str, None]`) to set "allow empty" to fields:

```python
class MyOtherModel(Model):
    baz: typing.Optional[str]


model = MyOtherModel()
model.validate()  # no exc raised
```

**Pull request merged**